File: src/mcp/__tests__/client.test.mjs
/**
 * Tests for GkillReadClient from gkill-read-server.mjs.
 *
 * All network calls are mocked via globalThis.fetch so no real server is needed.
 */

import { describe, test, expect, vi, beforeEach, afterEach } from "vitest";
import { GkillReadClient } from "../gkill-read-server.mjs";

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function mockFetchOk(body = {}) {
  globalThis.fetch = vi.fn().mockResolvedValue({
    ok: true,
    json: () => Promise.resolve(body),
  });
}

function mockFetchError(status, body = {}) {
  globalThis.fetch = vi.fn().mockResolvedValue({
    ok: false,
    status,
    json: () => Promise.resolve(body),
  });
}

// Save and restore env + fetch between tests.
let savedEnv;
let savedFetch;

beforeEach(() => {
  savedEnv = { ...process.env };
  savedFetch = globalThis.fetch;
  // Clear gkill-related env vars so defaults are used.
  delete process.env.GKILL_BASE_URL;
  delete process.env.GKILL_USER;
  delete process.env.GKILL_PASSWORD_SHA256;
  delete process.env.GKILL_PASSWORD;
  delete process.env.GKILL_LOCALE;
  delete process.env.GKILL_SESSION_ID;
  delete process.env.GKILL_INSECURE;
  delete process.env.GKILL_FETCH_TIMEOUT_MS;
});

afterEach(() => {
  process.env = savedEnv;
  globalThis.fetch = savedFetch;
});

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------
describe("GkillReadClient constructor", () => {
  test("sets defaults when env vars are absent", () => {
    const client = new GkillReadClient();
    expect(client.baseUrl).toBe("http://127.0.0.1:9999");
    expect(client.userId).toBe("");
    expect(client.passwordSha256).toBe("");
    expect(client.password).toBe("");
    expect(client.defaultLocale).toBe("ja");
    expect(client.sessionId).toBe("");
    expect(client.dispatcher).toBeNull();
  });

  test("reads env vars when set", () => {
    process.env.GKILL_BASE_URL = "https://example.com:1234";
    process.env.GKILL_USER = "testuser";
    process.env.GKILL_PASSWORD_SHA256 = "abc123";
    process.env.GKILL_LOCALE = "en";
    process.env.GKILL_SESSION_ID = "sess-42";

    const client = new GkillReadClient();
    expect(client.baseUrl).toBe("https://example.com:1234");
    expect(client.userId).toBe("testuser");
    expect(client.passwordSha256).toBe("abc123");
    expect(client.defaultLocale).toBe("en");
    expect(client.sessionId).toBe("sess-42");
  });
});

// ---------------------------------------------------------------------------
// buildApiUrl
// ---------------------------------------------------------------------------
describe("buildApiUrl", () => {
  test("constructs correct URL", () => {
    const client = new GkillReadClient();
    expect(client.buildApiUrl("/api/login")).toBe("http://127.0.0.1:9999/api/login");
  });

  test("constructs URL with custom base", () => {
    process.env.GKILL_BASE_URL = "https://gkill.example.com";
    const client = new GkillReadClient();
    expect(client.buildApiUrl("/api/get_kyous_mcp")).toBe(
      "https://gkill.example.com/api/get_kyous_mcp",
    );
  });
});

// ---------------------------------------------------------------------------
// resolvePasswordSha256
// ---------------------------------------------------------------------------
describe("resolvePasswordSha256", () => {
  test("returns GKILL_PASSWORD_SHA256 directly when set", () => {
    process.env.GKILL_PASSWORD_SHA256 = "deadbeef";
    const client = new GkillReadClient();
    expect(client.resolvePasswordSha256()).toBe("deadbeef");
  });

  test("computes SHA-256 from GKILL_PASSWORD when hash is not set", async () => {
    process.env.GKILL_PASSWORD = "secret";
    const client = new GkillReadClient();
    const hash = client.resolvePasswordSha256();
    // SHA-256 of "secret"
    const { createHash } = await import("node:crypto");
    const expected = createHash("sha256").update("secret").digest("hex");
    expect(hash).toBe(expected);
  });

  test("returns empty string when neither env var is set", () => {
    const client = new GkillReadClient();
    expect(client.resolvePasswordSha256()).toBe("");
  });
});

// ---------------------------------------------------------------------------
// hasErrors
// ---------------------------------------------------------------------------
describe("hasErrors", () => {
  test("returns true when errors array is non-empty", () => {
    const client = new GkillReadClient();
    expect(client.hasErrors({ errors: [{ error_code: "ERR" }] })).toBe(true);
  });

  test("returns false when errors array is empty", () => {
    const client = new GkillReadClient();
    expect(client.hasErrors({ errors: [] })).toBe(false);
  });

  test("returns false when errors field is missing", () => {
    const client = new GkillReadClient();
    expect(client.hasErrors({})).toBe(false);
  });

  test("returns false for null/undefined", () => {
    const client = new GkillReadClient();
    expect(client.hasErrors(null)).toBe(false);
    expect(client.hasErrors(undefined)).toBe(false);
  });
});

// ---------------------------------------------------------------------------
// login
// ---------------------------------------------------------------------------
describe("login", () => {
  test("calls fetch with correct URL and body", async () => {
    process.env.GKILL_USER = "admin";
    process.env.GKILL_PASSWORD_SHA256 = "hash123";
    mockFetchOk({ session_id: "test-session", errors: [] });

    const client = new GkillReadClient();
    const sessionId = await client.login();

    expect(sessionId).toBe("test-session");
    expect(globalThis.fetch).toHaveBeenCalledTimes(1);

    const [url, opts] = globalThis.fetch.mock.calls[0];
    expect(url).toBe("http://127.0.0.1:9999/api/login");
    expect(opts.method).toBe("POST");
    expect(JSON.parse(opts.body)).toEqual({
      user_id: "admin",
      password_sha256: "hash123",
      locale_name: "ja",
    });
  });

  test("returns existing sessionId without calling fetch", async () => {
    process.env.GKILL_SESSION_ID = "existing-session";
    mockFetchOk();

    const client = new GkillReadClient();
    const sessionId = await client.login();

    expect(sessionId).toBe("existing-session");
    expect(globalThis.fetch).not.toHaveBeenCalled();
  });

  test("throws when credentials are missing", async () => {
    const client = new GkillReadClient();
    await expect(client.login()).rejects.toThrow("Missing login credentials");
  });
});

// ---------------------------------------------------------------------------
// post
// ---------------------------------------------------------------------------
describe("post", () => {
  test("sends correct request and returns JSON body", async () => {
    const responseBody = { data: "result", errors: [] };
    mockFetchOk(responseBody);

    const client = new GkillReadClient();
    const result = await client.post("/api/test", { key: "value" });

    expect(result).toEqual(responseBody);
    expect(globalThis.fetch).toHaveBeenCalledTimes(1);

    const [url, opts] = globalThis.fetch.mock.calls[0];
    expect(url).toBe("http://127.0.0.1:9999/api/test");
    expect(opts.method).toBe("POST");
    expect(opts.headers["Content-Type"]).toBe("application/json");
    expect(JSON.parse(opts.body)).toEqual({ key: "value" });
  });

  test("throws GkillApiError on non-ok response", async () => {
    mockFetchError(500, { errors: [{ error_code: "ERR500" }] });

    const client = new GkillReadClient();
    await expect(client.post("/api/fail", {})).rejects.toThrow("HTTP 500");
  });

  test("throws GkillApiError on network failure", async () => {
    globalThis.fetch = vi.fn().mockRejectedValue(new Error("ECONNREFUSED"));

    const client = new GkillReadClient();
    await expect(client.post("/api/down", {})).rejects.toThrow("Network error");
  });
});

File: src/server/gkill/mcp/client_test.go
package mcp

// GkillClient（読み取りサーバの client）の検査。
//
// ネットワークは全部モック（http.Client の Transport 差し替え）なので実サーバは要らない。

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"net/http"
	"testing"
)

// ---------------------------------------------------------------------------
// Constructor
// ---------------------------------------------------------------------------
func TestGkillReadClientConstructor(t *testing.T) {
	t.Run("sets defaults when env vars are absent", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		expectEqual(t, client.BaseURL, "http://127.0.0.1:9999")
		expectEqual(t, client.UserID(), "")
		expectEqual(t, client.PasswordSha256, "")
		expectEqual(t, client.Password, "")
		expectEqual(t, client.DefaultLocale(), "ja")
		expectEqual(t, client.SessionID(), "")
		expectTrue(t, !client.Insecure, "insecure transport without GKILL_INSECURE")
	})

	t.Run("reads env vars when set", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_BASE_URL", "https://example.com:1234")
		t.Setenv("GKILL_USER", "testuser")
		t.Setenv("GKILL_PASSWORD_SHA256", "abc123")
		t.Setenv("GKILL_LOCALE", "en")
		t.Setenv("GKILL_SESSION_ID", "sess-42")

		client := NewGkillClientFromEnv()
		expectEqual(t, client.BaseURL, "https://example.com:1234")
		expectEqual(t, client.UserID(), "testuser")
		expectEqual(t, client.PasswordSha256, "abc123")
		expectEqual(t, client.DefaultLocale(), "en")
		expectEqual(t, client.SessionID(), "sess-42")
	})
}

// ---------------------------------------------------------------------------
// buildApiUrl
// ---------------------------------------------------------------------------
func TestBuildApiURL(t *testing.T) {
	t.Run("constructs correct URL", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		expectEqual(t, client.BuildApiURL("/api/login"), "http://127.0.0.1:9999/api/login")
	})

	t.Run("constructs URL with custom base", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_BASE_URL", "https://gkill.example.com")
		client := NewGkillClientFromEnv()
		expectEqual(t, client.BuildApiURL("/api/get_kyous_mcp"), "https://gkill.example.com/api/get_kyous_mcp")
	})
}

// ---------------------------------------------------------------------------
// resolvePasswordSha256
// ---------------------------------------------------------------------------
func TestResolvePasswordSha256(t *testing.T) {
	t.Run("returns GKILL_PASSWORD_SHA256 directly when set", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_PASSWORD_SHA256", "deadbeef")
		client := NewGkillClientFromEnv()
		expectEqual(t, client.ResolvePasswordSha256(), "deadbeef")
	})

	t.Run("computes SHA-256 from GKILL_PASSWORD when hash is not set", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_PASSWORD", "secret")
		client := NewGkillClientFromEnv()
		hash := client.ResolvePasswordSha256()
		// SHA-256 of "secret"
		sum := sha256.Sum256([]byte("secret"))
		expectEqual(t, hash, hex.EncodeToString(sum[:]))
	})

	t.Run("returns empty string when neither env var is set", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		expectEqual(t, client.ResolvePasswordSha256(), "")
	})
}

// ---------------------------------------------------------------------------
// hasErrors
// ---------------------------------------------------------------------------
func TestHasErrors(t *testing.T) {
	clearGkillEnv(t)
	client := NewGkillClientFromEnv()

	t.Run("returns true when errors array is non-empty", func(t *testing.T) {
		expectTrue(t, client.HasErrors(obj("errors", arr(obj("error_code", "ERR")))), "expected true")
	})

	t.Run("returns false when errors array is empty", func(t *testing.T) {
		expectTrue(t, !client.HasErrors(obj("errors", arr())), "expected false")
	})

	t.Run("returns false when errors field is missing", func(t *testing.T) {
		expectTrue(t, !client.HasErrors(obj()), "expected false")
	})

	t.Run("returns false for null/undefined", func(t *testing.T) {
		expectTrue(t, !client.HasErrors(nil), "expected false for nil")
	})
}

// ---------------------------------------------------------------------------
// login
// ---------------------------------------------------------------------------
func TestLogin(t *testing.T) {
	t.Run("calls fetch with correct URL and body", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_USER", "admin")
		t.Setenv("GKILL_PASSWORD_SHA256", "hash123")
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchOk(`{"session_id":"test-session","errors":[]}`)

		sessionID, err := client.Login(context.Background())
		expectNoError(t, err)
		expectEqual(t, sessionID, "test-session")
		expectTrue(t, len(fetch.calls) == 1, "fetch called %d times", len(fetch.calls))

		call := fetch.calls[0]
		expectEqual(t, call.URL, "http://127.0.0.1:9999/api/login")
		expectEqual(t, call.Method, "POST")
		expectEqual(t, parse(t, call.Body), obj(
			"user_id", "admin",
			"password_sha256", "hash123",
			"locale_name", "ja",
		))
	})

	t.Run("returns existing sessionId without calling fetch", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_SESSION_ID", "existing-session")
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchOk(`{}`)

		sessionID, err := client.Login(context.Background())
		expectNoError(t, err)
		expectEqual(t, sessionID, "existing-session")
		expectTrue(t, len(fetch.calls) == 0, "fetch was called")
	})

	t.Run("throws when credentials are missing", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		_, err := client.Login(context.Background())
		expectErrorContains(t, err, "Missing login credentials")
	})
}

// ---------------------------------------------------------------------------
// post
// ---------------------------------------------------------------------------
func TestPost(t *testing.T) {
	t.Run("sends correct request and returns JSON body", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchOk(`{"data":"result","errors":[]}`)

		result, err := client.Post(context.Background(), "/api/test", obj("key", "value"))
		expectNoError(t, err)
		expectEqual(t, result, obj("data", "result", "errors", arr()))
		expectTrue(t, len(fetch.calls) == 1, "fetch called %d times", len(fetch.calls))

		call := fetch.calls[0]
		expectEqual(t, call.URL, "http://127.0.0.1:9999/api/test")
		expectEqual(t, call.Method, "POST")
		expectEqual(t, call.Header.Get("Content-Type"), "application/json")
		expectEqual(t, parse(t, call.Body), obj("key", "value"))
	})

	// gkill は 2026-08 から異常時に 4xx/5xx を返す。ステータスだけで throw すると
	// 本文の errors が呼び出し側へ届かず、CallApi の再ログインが不通になる。
	t.Run("returns the body on non-ok response when it is a gkill envelope", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchError(500, `{"errors":[{"error_code":"ERR000410"}],"messages":null}`)

		result, err := client.Post(context.Background(), "/api/fail", obj())
		expectNoError(t, err)
		expectEqual(t, objAt(t, arrAt(t, result, "errors")[0]).Value("error_code"), "ERR000410")
	})

	t.Run("throws GkillApiError on non-ok response that is not a gkill envelope", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchError(502, `{"message":"Bad Gateway"}`)

		_, err := client.Post(context.Background(), "/api/fail", obj())
		expectGkillApiError(t, err)
		expectErrorContains(t, err, "HTTP 502")
	})

	// 本文が JSON でない非2xx (リバースプロキシの HTML 502/504、本文なしの 413/500 等) を
	// 「Failed to parse JSON response」にすると、HTTP ステータスが呼び出し側へ一切届かない。
	t.Run("includes the HTTP status when a non-2xx response body is not JSON", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.impl = func(_ int, _ fetchCall) (*http.Response, error) {
			return httpRaw(502, "text/html", []byte("<html>bad gateway</html>")), nil
		}

		_, err := client.Post(context.Background(), "/api/fail", obj())
		expectErrorContains(t, err, "HTTP 502")
	})

	t.Run("keeps the parse error when a 2xx response body is not JSON", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.impl = func(_ int, _ fetchCall) (*http.Response, error) {
			return httpRaw(200, "application/json", []byte("")), nil
		}

		_, err := client.Post(context.Background(), "/api/broken", obj())
		expectErrorContains(t, err, "Failed to parse JSON response")
	})

	t.Run("throws GkillApiError on network failure", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.mockFetchReject(errors.New("ECONNREFUSED"))

		_, err := client.Post(context.Background(), "/api/down", obj())
		expectGkillApiError(t, err)
		expectErrorContains(t, err, "Network error")
	})
}

// ---------------------------------------------------------------------------
// fetchFile
// ---------------------------------------------------------------------------
func TestFetchFile(t *testing.T) {
	// /files/ はエンベロープ(errors配列)を返さないので、セッション切れは HTTP 401 でしか
	// 分からない。CallApi と同じく、401 なら1回だけログインし直して再試行する。
	// MCP は長寿命プロセスなので、これが無いと期限切れ以降ファイル取得が復旧不能になる。
	t.Run("re-logins once and retries when the file request returns HTTP 401", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_USER", "admin")
		t.Setenv("GKILL_PASSWORD_SHA256", "hash")
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.impl = func(callCount int, _ fetchCall) (*http.Response, error) {
			switch callCount {
			case 1:
				// First file request with the expired session
				return httpRaw(401, "", nil), nil
			case 2:
				// Re-login
				return httpJSON(200, `{"session_id":"new-sess","errors":null}`), nil
			}
			// Retry succeeds
			return httpRaw(200, "image/png", make([]byte, 4)), nil
		}

		result, err := client.FetchFile(context.Background(), "/files/rep/photo.png", "expired-sess")
		expectNoError(t, err)
		expectEqual(t, result.ContentType, "image/png")
		expectTrue(t, len(result.Buffer) == 4, "buffer has %d bytes", len(result.Buffer))
		expectTrue(t, len(fetch.calls) == 3, "fetch called %d times", len(fetch.calls))

		// 再試行はログインし直したセッションの Cookie で行われること
		mustContain(t, fetch.calls[2].Header.Get("Cookie"), "new-sess")
	})

	t.Run("throws when the retry after re-login still returns HTTP 401", func(t *testing.T) {
		clearGkillEnv(t)
		t.Setenv("GKILL_USER", "admin")
		t.Setenv("GKILL_PASSWORD_SHA256", "hash")
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.impl = func(callCount int, _ fetchCall) (*http.Response, error) {
			if callCount == 2 {
				// Re-login
				return httpJSON(200, `{"session_id":"new-sess","errors":null}`), nil
			}
			// First attempt and the retry both get 401
			return httpRaw(401, "", nil), nil
		}

		_, err := client.FetchFile(context.Background(), "/files/rep/photo.png", "expired-sess")
		expectErrorContains(t, err, "HTTP 401 fetching file")
		expectTrue(t, len(fetch.calls) == 3, "fetch called %d times", len(fetch.calls))
	})

	// 401 以外は再ログインしない。read_handlers の 404→原因名指しの変換は
	// 生の「HTTP 404 fetching file」がそのまま1回で返ることに依存している。
	t.Run("throws immediately without re-login on non-401 errors", func(t *testing.T) {
		clearGkillEnv(t)
		client := NewGkillClientFromEnv()
		fetch := installFetchMock(client)
		fetch.impl = func(_ int, _ fetchCall) (*http.Response, error) { return httpRaw(404, "", nil), nil }

		_, err := client.FetchFile(context.Background(), "/files/rep/missing.png", "sess")
		expectErrorContains(t, err, "HTTP 404 fetching file")
		expectTrue(t, len(fetch.calls) == 1, "fetch called %d times", len(fetch.calls))
	})
}
